front/src/applications/operationalStudies/components/MacroEditor/OsrdNgeSync.tsx

```diff
@@ -0,0 +1,918 @@
+      const fixed = nodeIndexed!.saved ? 1 : 0;
+      return {
+        key,
+        x: nodeIndexed!.node.position_x,
+        y: nodeIndexed!.node.position_y,
+        height: 100,
+        width: 50,
+        fixed,
+      };
+    });
+
+    const nodeIndexById = nodes.reduce(
+      (acc, curr, index) => {
+        acc[curr.key] = index;
+        return acc;
+      },
+      {} as Record<string, number>
+    );
+
+    const edges: { source: number; target: number; weight: number }[] = [];
+    const edgesIndex: Record<string, number> = {};
+    this.trainSchedules.forEach((ts) => {
+      let prevInNodeArrayIndex: number | null = null;
+      ts.path.forEach((item) => {
+        // create edges
+        const targetKey = OsrdNgeSync.getPathKey(item);
+        const targetNodeIndexInNodeArray = nodeIndexById[targetKey];
+        if (prevInNodeArrayIndex) {
+          const edgeIndexKey = `${prevInNodeArrayIndex}-${targetNodeIndexInNodeArray}`;
+          if (edgesIndex[edgeIndexKey]) {
+            edges[edgesIndex[edgeIndexKey]].weight += 1;
+          } else {
+            const newLength = edges.push({
+              source: prevInNodeArrayIndex,
+              target: targetNodeIndexInNodeArray,
+              weight: 1,
+            });
+            edgesIndex[edgeIndexKey] = newLength - 1;
+          }
+        }
+        prevInNodeArrayIndex = targetNodeIndexInNodeArray;
+      });
+    });
+
+    return {
+      nodes,
+      edges,
+    };
+  }
+
+  /**
+   * Apply a layout on nodes and save the new position.
+   * Nodes that are saved are fixed.
+   */
+  private applyLayout() {
+    const graph = this.toGraph();
+    if (graph.nodes.some((n) => !n.fixed)) {
+      const layout = new Layout()
+        .defaultNodeSize(100)
+        .size([800, 600])
+        .linkDistance(200)
+        .avoidOverlaps(true)
+        .nodes(graph.nodes)
+        .links(graph.edges)
+        .start(5, 10, 15, 20, false, false);
+
+      layout.nodes().forEach((n) => {
+        const node = n as ColaNode & { key: string };
+        this.updateNodeDataByKey(node.key, {
+          position_x: Math.round(node.x),
+          position_y: Math.round(node.y),
+        });
+      });
+    }
+  }
+
+  /**
+   * Get nodes of the scenario that are saved in the DB.
+   */
+  private async apiGetSavedNodes(): Promise<MacroNodeResponse[]> {
+    const pageSize = 100;
+    let page = 1;
+    let reachEnd = false;
+    const result: MacroNodeResponse[] = [];
+    while (!reachEnd) {
+      const promise = this.dispatch(
+        osrdEditoastApi.endpoints.getProjectsByProjectIdStudiesAndStudyIdScenariosScenarioIdMacroNodes.initiate(
+          {
+            projectId: this.scenario.project.id,
+            studyId: this.scenario.study_id,
+            scenarioId: this.scenario.id,
+            pageSize,
+            page,
+          }
+        )
+      );
+      // need to unsubscribe on get call to avoid cache issue
+      promise.unsubscribe();
```

Comment:
Nit: would be cleaner to use `{ subscribe: false }`

Reply:
I'm not sure that `{subscribe: false }` disable the cache.
I tried many patterns to use RTK outside a react component, and finally I found this working pattern (which is one given in the documentation : https://redux-toolkit.js.org/rtk-query/usage/usage-without-react-hooks#adding-a-subscription).

Reply:
The pattern described in the docs is to unsubscribe some time after making the request. To not subscribe at all, the options can be used.